src/ShallowWrapper.js

```diff
@@ -1078,13 +1100,14 @@ class ShallowWrapper {
   dive(options = {}) {
     const name = 'dive';
     return this.single(name, (n) => {
-      if (isDOMComponentElement(n)) {
+      if (n && n.nodeType === 'host') {
         throw new TypeError(`ShallowWrapper::${name}() can not be called on DOM components`);
```

Comment:
I'm just going to change the terminology to "Host Components". This PR moves enzyme to an adapter architecture, but removing concepts of react entirely from enzyme and making it possible to use enzyme for react-like libraries is *close*, but will require some additional refactoring i'm sure. For things like terminology of "DOM Components" etc, I think these ideas are fairly baked in in other parts of the code as well. I think that that's okay at this point.